### ISSUE TITLE
Do we really need profile sections?

### ISSUE THREAD

**Issue by felipesanches** [googlefonts/fontbakery#3336]
As far as I can tell, none of the existing FontBakery profiles in this repo use more than a single section. It is also not clear to me what benefit we would get from having sections at all.

@graphicore, does the current implementation of profile sections have any interesting/useful feature for this case of multiple sections? I've been reading those portions of the source code but I could not yet figure it out. If not, do we have plans to implement something that uses the info about checks being grouped in sections? Otherwise, I think it only makes the codebase more complex than necessary and I would advocate for deprecating sections.

To be honest, I feel that sections, as a mere subset of a profile, seem to be almost redundant with the concept of a profile. Also, if one needs to run only a subset of a profile, one can use the -c command line flag with keywords that are present in the IDs of the multiple checks needed.

**Comment by graphicore** [comment]
I'd have to look into it more deeply, but in general you are probably right: we failed to use them to create reporting tools for users that help them organizing the information of a report by e.g. section.

In defense of sections, we actually use them when we import profiles into other profiles, to group the checks from a profile into a section, so when different profiles end up in a so called vendor profile, we can look at the section of a check and know from which profile it's coming. The `-S`,`--show-sections` flag actually creates a per section summary, like this excerpt of a  command line report:

```
======== Section results: <Section: fontbakery.profiles.loca> ========
1 check in section

    ERROR: 0
    FAIL: 0
    WARN: 0
    INFO: 0
    SKIP: 0
    PASS: 1

======== Section results: <Section: fontbakery.profiles.hhea> ========
2 checks in section

    ERROR: 0
    FAIL: 0
    WARN: 0
    INFO: 0
    SKIP: 0
    PASS: 2

======== Section results: <Section: fontbakery.profiles.dsig> ========
1 check in section

    ERROR: 0
    FAIL: 1
    WARN: 0
    INFO: 0
    SKIP: 0
    PASS: 0
```

In so far, there's a relation between single-purpose profiles and sections AND you are wrong saying that not a single profile uses them. What you observe is that not a single profile creates explicitly multiple sections. We also had a discussion once to get rid of all profiles altogether, not sure if this is related. The original profile, e.g. for grouping or for filtering, could also be a property of the check.

**Comment by felipesanches** [comment]
> In defense of sections, we actually use them when we import profiles into other profiles, to group the checks from a profile into a section, so when different profiles end up in a so called vendor profile, we can look at the section of a check and know from which profile it's coming. The `-S`,`--show-sections` flag actually creates a per section summary, like this excerpt of a command line report:
> ...

I see. But I don't see much value in this reporting of check-results by opentype table sections. I would be OK with deprecating that feature in order to make the codebase less complex.

> We also had a discussion once to get rid of all profiles altogether, not sure if this is related.
Not related at all.

If I recall correctly it was not a proposal to get rid of profiles, but instead a discussion about being able to invoke checks by their check-id without having to also specify the profile. But let's not talk about that here, because that proposal is completely unrelated to what we're discussing here.

**Comment by felipesanches** [comment]
My conclusion is that I want to completely remove sections from FontBakery. Do you have any advice for things to consider while doing so?

**Comment by graphicore** [comment]
> I see. But I don't see much value in this reporting of check-results by opentype table sections.

We should rather ask actual users if such a report is/would be useful to them. But, I can imagine that it's not a well known feature.

It can help you gather information in what general area your problems are located.

The fact that our profiles are by opentype table has nothing to do with the feature, it's a downstream profile authoring decision. 

> I would be OK with deprecating that feature in order to make the codebase less complex.

Yeah, but this thing is not **that** complex. Whatever.

> My conclusion is that I want to completely remove sections from FontBakery. Do you have any advice for things to consider while doing so?

If you really do it, make it a major version change.

And don't break stuff. :laughing: It will break stuff for sure. If I would write FB now, I wouldn't add sections because we're not really using them. I still think it can be useful and I believe other similar tools (e.g. unit testing frameworks) have similar grouping features, so somewhere someone decided it would be a nice to have feature.

**Comment by felipesanches** [comment]
> > I see. But I don't see much value in this reporting of check-results by opentype table sections.
> 
> We should rather ask actual users if such a report is/would be useful to them. But, I can imagine that it's not a well known feature.

Agreed. Here are the things that I see as evidence that this feature is probably not used by anybody:
- we do not use it extensively ourselves (which tends to work as an example and people frequently replicate and adapt things they already see working somewhere else)
- it is a feature I do not remember ever talking about or hearing other people talk about at all.
- we did not get bugfix PRs or issues regarding this feature, which is something that typically happens whenever someone uses a feature, even if doing so privately

> It can help you gather information in what general area your problems are located.

yes, maybe. But that could also be just wishful thinking.

> The fact that our profiles are by opentype table has nothing to do with the feature, it's a downstream profile authoring decision.
Sure!

> > I would be OK with deprecating that feature in order to make the codebase less complex.
> 
> Yeah, but this thing is not **that** complex. Whatever.

Legacy code is code that needs to be maintained. It gets in the way when trying to implement new stuff. It got in my way yesterday while I was working on another issue (#3139), for instance.

> > My conclusion is that I want to completely remove sections from FontBakery. Do you have any advice for things to consider while doing so?
> 
> If you really do it, make it a major version change.

YES! That's my plan.

> 
> And don't break stuff. laughing It will break stuff for sure.

If somebody is using that, we'll know. And they'll have to adapt. And, again, I think it is extremely unlikely that anyone is using this feature at all.

> If I would write FB now, I wouldn't add sections because we're not really using them. I still think it can be useful and I believe other similar tools (e.g. unit testing frameworks) have similar grouping features, so somewhere someone decided it would be a nice to have feature.

I can see that this feature might be useful if we were actively using it. But we're not. And it makes the code a bit harder to maintain & improve. Maybe I am just dealing with a destructive urge in order to clean up that portion of the code. I see no problem if we decide to reintroduce the feature in the future.

### “The urge for destruction is also a creative urge!”

**Comment by graphicore** [comment]
> * we did not get bugfix PRs or issues regarding this feature, which is something that typically happens whenever someone uses a feature, even if doing so privately

Yes, but in general, the checkrunner related code is pretty stable since many years and *everyone uses sections all the time*, yet probably not explicitly.

>Legacy code is code that needs to be maintained. It gets in the way when trying to implement new stuff. It got in my way yesterday while I was working on another issue (#3139), for instance. 

I commented there, I don't think that example is related. I don't see how this is a maintenance burden. Also, I'd be available to help maintain stuff in that region.

>  And they'll have to adapt. 

I may at some point have to fork Font Bakery, the checkrunner, not the font specific profiles. I see how some people here want to ultra focus Font Bakery on just the current narrow use case and I disagree with the idea and reasoning around it. I rather believe that a tool that is a bit more generic/flexible than the most narrow current use case is valuable; also because it keeps from introducing bad habits e.g. it helps preserving a separation of concerns.

>>    If I would write FB now, I wouldn't add sections because we're not really using them. I still think it can be useful and I believe other similar tools (e.g. unit testing frameworks) have similar grouping features, so somewhere someone decided it would be a nice to have feature.
>
>I can see that this feature might be useful if we were actively using it. But we're not. And it makes the code a bit harder to maintain & improve. 

I disagree a bit. It's super unproblematic code and it's stable since years.

> Maybe I am just dealing with a destructive urge in order to clean up that portion of the code.

Sounds as if it was bad and needed cleaning up, but it is not much code and pretty straight forward simple stuff.

>“The urge for destruction is also a creative urge!”

That's not reasoning, it's just a platitude.

**Comment by jmsole** [comment]
Hi. I hope you decide not to get rid of Sections. I think a big issue right now is that it's not clear how to use them on a custom profile. I would like to create my own profile and reorganise imported checks into sections that make sense to me, rather than have section names that are less meaningful. I don't care if a check comes from the Notofonts profile, I care about it being grouped with other checks that operate in similar areas.

Would it be possible for someone to provide a small example of how Sections could be used to achieve what I describe above?

Cheers, and thanks in advance,
José

**Comment by PeterDekkers** [comment]
Hello @jmsole. I agree with you, sections are useful for providing custom groupings within a project's specific context.

Here is a basic example that sets up custom sections. Note though, I have been unable to _order_ the check results within the report. I've tried various combinations of passing `order` or `custom_order` arguments, but none seem to work. And the documentation on these parameters is basic—I can't really make sense of it.

```python
from fontbakery.fonts_profile import profile_factory
from fontbakery.profiles.universal import (
    profile as profile_universal,
    DESIGNSPACE_CHECKS,
)
from fontbakery.profiles.ufo_sources import (
    profile as profile_ufo_sources,
)

profile = profile_factory(
    sections=[
        Section(
            "UFO linting & fields",
            checks=profile_ufo_sources.checks,
        ),
        Section(
            "UFO designspace",
            # Limit to only `DESIGNSPACE_CHECKS`
            checks=[
                check
                for check in profile_universal.checks
                if check.id in DESIGNSPACE_CHECKS
            ],
        ),
    ],
)
```

**Comment by felipesanches** [comment]
Reorganizing the existing checks into meaningful sections would be nice, but nobody has put effort into it, as far as I am aware. That would be a very nice contribution if anyone reading this feels like participating in the project!

**Comment by PeterDekkers** [comment]
Great, I'd be happy to give that a go @felipesanches.

Any immediate ideas from any of the maintainers as to how to approach this would be welcome.

**Comment by PeterDekkers** [comment]
I suppose maintaining the order in which checks are added to a section would be a nice start.

**Comment by PeterDekkers** [comment]
@simoncozens I have a question for you, if you don't mind. The order of the checks as they are added to Sections is lost because of [this line here](https://github.com/googlefonts/fontbakery/blob/ec1103221206d0e2c5394e827c5d175da6250385/Lib/fontbakery/profile.py#L625), which is part of a commit of yours.

To summarise, the line in `profile.py` does:
```python
scopes = self._analyze_checks(full_order, checks)
# check, signature, scope = item
scopes.sort(key=lambda item: item[1], reverse=reverse)
```

This sorts the checks on a `signature` tuple, but in doing so, loses their original order. The signature is determined (in [_analyze_checks](https://github.com/googlefonts/fontbakery/blob/ec1103221206d0e2c5394e827c5d175da6250385/Lib/fontbakery/profile.py#L461)) based on the number of arguments passed to the check (if I understand it correctly).

What's the reasoning behind the signature, if you recall? Thank you.

**Comment by simoncozens** [comment]
Don't look at me - that code has actually been there for ever!
https://github.com/googlefonts/fontbakery/commit/1e8217dea3195d1760facef80361f66c5e9704fe

**Comment by PeterDekkers** [comment]
@graphicore, you're the next one down the commit chain. Any input on what this mechanism (two posts up) is for?

Thanks so much! I don't want to override it if it's important.

**Comment by graphicore** [comment]
The first thing I notice is that it was possible to pass a `key` function to that sort in https://github.com/googlefonts/fontbakery/commit/1e8217dea3195d1760facef80361f66c5e9704fe but it is no more in since  https://github.com/googlefonts/fontbakery/commit/17434547df4f28358d5be33eb6a737364348222f both committed on the same day, so the refactoring was probably going on for some days before I made a bulk of commits.

From https://docs.python.org/3/library/stdtypes.html#list.sort

> key specifies a function of one argument that is used to extract a comparison key from each list element (for example, key=str.lower). The key corresponding to each item in the list is calculated once and then used for the entire sorting process. The default value of None means that list items are sorted directly without calculating a separate key value.

and

>The [sort()](https://docs.python.org/3/library/stdtypes.html#list.sort) method is guaranteed to be stable. A sort is stable if it guarantees not to change the relative order of elements that compare equal — this is helpful for sorting in multiple passes (for example, sort by department, then by salary grade).

In the older case, passing e.g. `key=lambda item: 1`, should keep any order in tact. I'm not sure why it was removed by me. The reason is likely, that it was never used.


> What's the reasoning behind sorting on the signature—and behind _analyze_checks in general—if you recall? Thank you.

The main reasoning on sorting is that there's eventually a defined order of checks execution, which is mostly important when producing an ordered output, as for the terminal logger, which is linear in nature. As it turns out there are different ways in that such an order could be grouped, mainly in the most used simple case:
* (A) first fonts then checks: this would execute/group by the first font and run all checks with it, then by the second font and run all checks with it etc.
* (B) first checks then fonts: this would execute/group by the first check then run it with all fonts, then by the second check and run it with all fonts, etc.

There's an ` -o ORDER, --order ORDER` argument:

```
                        Comma separated list of order arguments.
                        The execution order is determined by the order of the check
                        definitions and by the order of the iterable arguments.
                        A section defines its own order. `--order` can be used to
                        override the order of *all* sections.
                        Despite the ITERATED_ARGS there are two special
                        values available:
                        "*iterargs" -- all remainig ITERATED_ARGS
                        "*check"    -- order by check
                        ITERATED_ARGS: designspace, font, glyphs_file, ufo
                        A sections default is equivalent to: "*iterargs, *check".
                        A common use case is `-o "*check"` when checking the whole 
                        collection against a selection of checks picked with `--checkid`.
```
As you can see the default is `*iterargs, *check"` which is like the (A) case above i.e. first use all of the iterargs, which in a simple case is just `font`. "iterargs" is the args that are iterated to checking one font after another. As a check with the signature `(font)` will run once for each font.

When you specify `-o *check` the execution order will be by check first as in the (B).

However, `-o check` didn't work for me either and by looking into it I found a bug, which is caused by  switching from python2 to python3 PR: https://github.com/googlefonts/fontbakery/pull/4132 In the argument parsing, the order argument is filtered using `filter` which in py2 would result in a list but in py results in an iterator. That iterator is exhausted by the first section that is processed for order and subsequently order is just the the default for the remaining sections.

Sorting on the signature is in so far relevant, as by changing the `--order` argument, the signature is changed as well. When *check is first, the first element in the signature is always `1`, hence keeping the order of checks stable, but sorting by subsequent arguments. If subsequently, there's a `font` in the arguments, when it's used by the check, it's value is 1 otherwise it's 0 (font is not used), that would change the sort order.

`_analyze_checks` also keeps track of `iterargs` which are introduced indirectly via `@conditions` used by the check, so if an `@condition` uses font, the check is still executed once for each `font`, which is interpreted by the condition.

**Comment by graphicore** [comment]
The above works for me in the terminal reporter, but I don't know well Sections are treated in other reporters. Also in terminal reporter with `-j` the output order will be scrambled. That and the added information for section start and section end which was making the report harder to read led to removing that information from the terminal reporter. Sections were treated "stepmotherly"  in the past, so some of the reporting for it lacks behind and I believe some of the tooling may have been removed as well.

Just  for the shaping section (which produces only SKIP results with my setup):

 I removed some output information and grouped optically to highlight the subject.
Note how each check is executed  for each font before going on to the next check.

```
fonts/ofl/abeezee$ fontbakery check-googlefonts -n -o *check --succinct  -l PASS *

 >> Check:   com.google.fonts/check/shaping/regression
    Files:   ABeeZee-Italic.ttf
 >> Check:   com.google.fonts/check/shaping/regression
    Files:   ABeeZee-Regular.ttf


 >> Check:   com.google.fonts/check/shaping/forbidden
    Files:   ABeeZee-Italic.ttf
 >> Check:   com.google.fonts/check/shaping/forbidden
    Files:   ABeeZee-Regular.ttf


 >> Check:   com.google.fonts/check/shaping/collides
    Files:   ABeeZee-Italic.ttf
 >> Check:   com.google.fonts/check/shaping/collides
    Files:   ABeeZee-Regular.ttf
```

Note how for each font all checks are executed going on to the next font.

```
fonts/ofl/abeezee$ fontbakery check-googlefonts -n  --succinct  -l PASS *

 >> Check:   com.google.fonts/check/shaping/regression
    Files:   ABeeZee-Italic.ttf
 >> Check:   com.google.fonts/check/shaping/forbidden
    Files:   ABeeZee-Italic.ttf
>> Check:   com.google.fonts/check/shaping/collides
    Files:   ABeeZee-Italic.ttf

 >> Check:   com.google.fonts/check/shaping/regression
    Files:   ABeeZee-Regular.ttf
 >> Check:   com.google.fonts/check/shaping/forbidden
    Files:   ABeeZee-Regular.ttf
 >> Check:   com.google.fonts/check/shaping/collides
    Files:   ABeeZee-Regular.ttf
```

Use -S to get some information about the sections (each profile creates a section currently).

```

======== Section results: <Section: Google Fonts> ========
270 checks in section

    ERROR: 0
    FAIL: 10
    WARN: 16
    INFO: 11
    SKIP: 54
    PASS: 179

======== Section results: <Section: Universal> ========
64 checks in section

    ERROR: 0
    FAIL: 0
    WARN: 10
    INFO: 4
    SKIP: 10
    PASS: 40

======== Section results: <Section: fontbakery.profiles.cff> ========
6 checks in section

    ERROR: 0
    FAIL: 0
    WARN: 0
    INFO: 0
    SKIP: 6
    PASS: 0

======== Section results: <Section: fontbakery.profiles.cmap> ========
1 check in section

    ERROR: 0
    FAIL: 0
    WARN: 0
    INFO: 0
    SKIP: 0
    PASS: 1

======== Section results: <Section: fontbakery.profiles.head> ========
7 checks in section

    ERROR: 0
    FAIL: 0
    WARN: 0
    INFO: 0
    SKIP: 2
    PASS: 5

======== Section results: <Section: fontbakery.profiles.os2> ========
13 checks in section

    ERROR: 0
    FAIL: 0
    WARN: 0
    INFO: 0
    SKIP: 2
    PASS: 11

...
```

**Comment by PeterDekkers** [comment]
Thank you @graphicore for all the detailed background information. I will process it all and see if I can come up with some sensible—and, ideally, backwards-compatible—improvements to customise check grouping and sorting (for both runs and reports).
